### PR TITLE
feat(api): Complete email preview logic

### PR DESCRIPTION
**Pull request by tatarco**
### What changed? Why was the change needed?
Completed the work around maily hydration and expansion, including modifying validation and preparation for payload schema restructuring 

### Screenshots

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: .source
Comment (djabarovgeorge) [inline .source:1]: if we do not have ee changes in this pr, we should not have this hash changes

Comment (SokratisVidros) [reply .source:1]: Correct, this needs to be reverted.☝️

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts [hunk 1]
```diff
@@ -2,76 +2,146 @@
...
   }
 
-  private processItemNode(node: TipTapNode, item: any): TipTapNode {
-    if (node.type === 'text' && typeof node.text === 'string') {
-      const regex = /{#item\.(\w+)#}/g;
-      node.text = node.text.replace(regex, (_, key: string) => {
-        const propertyName = key;
-
-        return item[propertyName] !== undefined ? item[propertyName] : _;
+  private traverseAndAugment(node: TipTapNode, parentNode?: TipTapNode) {
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts]: Nitpicky comment to improve readability and self-documentation

```suggestion
  private processShowAndForControls(node: TipTapNode, parentNode?: TipTapNode) {
```

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts [hunk 2]
```diff
@@ -2,76 +2,146 @@
...
+    } else if (this.hasEach(node)) {
+      const newContent = this.expendedForEach(node);
+      node.content = newContent;
+      this.removeForNodeAndUpgradeContent(node, newContent, parentNode);
+    }
+  }
 
-    if (node.content) {
-      node.content = node.content.map((innerNode) => this.processItemNode(innerNode, item));
+  private removeForNodeAndUpgradeContent(node: TipTapNode, expandedContents: TipTapNode[], parentNode?: TipTapNode) {
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts]: Another one: I would move this code into `processShowAndForControls` and keep all that processing in a single function as a single unit of work.

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts [hunk 3]
```diff
@@ -2,76 +2,146 @@
...
+      const hydratedContent = this.replacePlaceholders(templateContent, value);
+      expandedContent.push(...hydratedContent);
     }
 
-    if (this.hasEachAttr(node)) {
-      return { type: 'section', content: this.handleFor(node) };
+    return expandedContent;
+  }
+
+  private isOrderedList(templateContent: TipTapNode[]) {
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts]: Please add some comments about the extra handling we do here for bullets and ordered list and why it's necessary.

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-output-renderer.usecase.ts [hunk 4]
```diff
@@ -1,44 +1,49 @@
 import { EmailRenderOutput, TipTapNode } from '@novu/shared';
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/email-output-renderer.usecase.ts]: Our use cases are named with verbs. So this one should be `render-email-output.usecase.ts`. I just saw that the names are also off for other steps. We can fix them in a follow up PR, it's not urgent.

File: apps/api/src/exception-filter.ts [hunk 5]
```diff
@@ -133,10 +147,34 @@ function handleZod(exception: ZodError<any>) {
 
   return { status, message };
 }
+function handleMongo(exception: MongoServerError) {
```
Comment (SokratisVidros) [inline apps/api/src/exception-filter.ts]: Is this change relevant to the email preview logic? If no it should have been a separate PR.

File: apps/api/src/app/workflows-v2/usecases/placeholder-enrichment/payload-preview-value-generator.usecase.ts [hunk 6]
```diff
@@ -22,20 +18,29 @@ export class CreateMockPayloadUseCase {
...
-  private buildPayloadForEmailEditor(controlValue: unknown): Record<string, unknown> {
-    const collectPlaceholderMappings = this.collectPlaceholdersFromTipTapSchemaUsecase.execute({
-      node: controlValue as TipTapNode,
-    });
-    const transformPlaceholderMap = this.transformPlaceholderMapUseCase.execute({ input: collectPlaceholderMappings });
+  private safeAttemptToParseEmailSchema(controlValue: string) {
+    try {
+      const { nestedPayload } = this.hydrateEmailSchemaUseCase.execute({
+        emailEditor: controlValue,
+        masterPayload: {
```
Comment (SokratisVidros) [inline apps/api/src/app/workflows-v2/usecases/placeholder-enrichment/payload-preview-value-generator.usecase.ts]: This concept of `masterPayload` is a bit confusing as it makes the reader believe there are different types of payloads in the app. How about passing payload, subscriber, and steps directly as arguments?

File: apps/api/src/app/environments-v1/usecases/construct-framework-workflow/construct-framework-workflow.usecase.ts [hunk 7]
```diff
@@ -21,6 +21,12 @@ import {
   SmsOutputRendererUsecase,
 } from '../output-renderers';
 
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface MasterPayload {
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/construct-framework-workflow/construct-framework-workflow.usecase.ts]: See my comment later about introducing this new term of MasterPayload.

File: apps/api/src/app/environments-v1/usecases/output-renderers/expend-email-editor-schema-command.ts [hunk 8]
```diff
@@ -3,7 +3,6 @@
 import { BaseCommand } from '@novu/application-generic';
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/expend-email-editor-schema-command.ts]: Please rename to `expand-email-editor-schema.command.ts`. There is a typo in the current filename.

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts [hunk 9]
```diff
@@ -2,76 +2,146 @@
 import { TipTapNode } from '@novu/shared';
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts]: Ditto about renaming the use case to `expand-email-schema.usecase.ts`

File: apps/api/src/app/environments-v1/usecases/output-renderers/hydrate-email-schema.usecase.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
+/* eslint-disable no-param-reassign */
+import { Injectable } from '@nestjs/common';
+import { TipTapNode } from '@novu/shared';
+import { z } from 'zod';
+import { MasterPayload } from '../construct-framework-workflow';
+import { HydrateEmailSchemaCommand } from './hydrate-email-schema.command';
+
+@Injectable()
+export class HydrateEmailSchemaUseCase {
```
Comment (SokratisVidros) [inline apps/api/src/app/environments-v1/usecases/output-renderers/hydrate-email-schema.usecase.ts:8]: One thing that sounds a bit confusing is why do we have hydrate and expand email use cases separately. I'd expect a single use case that does both.

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: apps/api/src/app/workflows-v2/usecases/generate-preview/construct-payload-from-placeholders-with-defaults-use-case.service.ts
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/generate-preview/construct-payload-from-placeholders-with-defaults-use-case.service.ts:1]: can we find a better location for this file? The route folder should be kept clean, containing only module file, controller, and folders.

File: packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts [hunk 11]
```diff
@@ -59,8 +59,14 @@ export class ControlPreviewIssue {
   variableName?: string;
   message: string;
 }
+export class MasterPayload {
```
Comment (djabarovgeorge) [inline packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts]: what do you think about changing the name here to something else?

```suggestion
export class MainPayload {
```

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts (hunk 9 above)
Comment (SokratisVidros) [reply apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts]: Reminder for this.

File: packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts (hunk 11 above)
Comment (SokratisVidros) [reply packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts]: I agree. The name is too generic and introduces a master payload concept that doesn't exist in the product. As said earlier, we should eliminate it https://github.com/novuhq/novu/pull/6772/files#r1827360737

File: apps/api/src/app/workflows-v2/usecases/placeholder-enrichment/payload-preview-value-generator.usecase.ts (hunk 6 above)
Comment (tatarco) [reply apps/api/src/app/workflows-v2/usecases/placeholder-enrichment/payload-preview-value-generator.usecase.ts]: just because there will be more, we havn't done actor yet,. that's the reason I added this concept, because its a higher level payload that includes all not only the customer provided payload, I want to minimize the changes when we add actor and others to the framework, create a structure that minimizes the maintenance, 
eventually the placeholders are not reffering to payload, they reffer to such a master payload.

File: packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts (hunk 11 above)
Comment (tatarco) [reply packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts]: is main payload exists?

Comment (SokratisVidros) [reply packages/shared/src/dto/step-schemas/generate-preview-response.dto.ts]: This is just a grouping for which we don't have a strong argument yet. I suggest we use separate parameters for payload, subscriber, and steps, as they all have a very specific meaning in Novu land.

If we need a new umbrella term, we will add it in the future in a more structured way.

File: apps/api/src/exception-filter.ts (hunk 5 above)
Comment (tatarco) [reply apps/api/src/exception-filter.ts]: i'll revert it

File: apps/api/src/app/environments-v1/usecases/construct-framework-workflow/construct-framework-workflow.usecase.ts (hunk 7 above)
Comment (tatarco) [reply apps/api/src/app/environments-v1/usecases/construct-framework-workflow/construct-framework-workflow.usecase.ts]: lets talk about it, I would like to find another solution but having to maintain all seperatly passing them in and out of functions, we should have a term for the all inclusive payload including us and the customers

File: apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts (hunk 9 above)
Comment (tatarco) [reply apps/api/src/app/environments-v1/usecases/output-renderers/email-schema-expander.usecase.ts]: all good, i havn't seen the comments before now, all being handled

File: apps/api/src/app/environments-v1/usecases/output-renderers/hydrate-email-schema.usecase.ts (hunk 10 above)
Comment (tatarco) [reply apps/api/src/app/environments-v1/usecases/output-renderers/hydrate-email-schema.usecase.ts:8]: I'm trying to create clear maintainable logic. there are two distinct behaviors here, we hydrate (from pointers to data) and than perform logic on data, those are hard to follow algorithms so I don't want to create a job security algorithem but one which is easy to learn, I'll add a comment explaining it

File: apps/api/src/app/workflows-v2/usecases/generate-preview/construct-payload-from-placeholders-with-defaults.usecase.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
+/* eslint-disable no-param-reassign */
```
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/generate-preview/construct-payload-from-placeholders-with-defaults.usecase.ts]: why do we need this?

Comment (tatarco) [reply apps/api/src/app/workflows-v2/usecases/generate-preview/construct-payload-from-placeholders-with-defaults.usecase.ts]: there is a lint not allowing reassignment, but it's used in this logic

**Review by SokratisVidros — APPROVED** [review]
